core/domain/opportunity_services.py

```diff
@@ -0,0 +1,370 @@
+        exp_opportunity_models)
+
+
+def update_exploration_opportunities(old_story, new_story):
+    """Updates the opportunities models according to the changes made in the
+    story.
+
+    Args:
+        old_story: Story. The old story object which is now updated.
+        new_story: Story. The new story object.
+    """
+    model_ids_need_update = set([])
+    exp_ids_in_old_story = old_story.story_contents.get_all_linked_exp_ids()
+    exp_ids_in_new_story = new_story.story_contents.get_all_linked_exp_ids()
+    new_added_exp_ids = set(exp_ids_in_new_story) - set(exp_ids_in_old_story)
+    deleted_exp_ids = set(exp_ids_in_old_story) - set(exp_ids_in_new_story)
+    unchanged_exp_ids = set(exp_ids_in_new_story) - new_added_exp_ids
+    if old_story.title != new_story.title:
+        model_ids_need_update |= set(unchanged_exp_ids)
+    else:
+        for exp_id in unchanged_exp_ids:
+            new_node = (
+                new_story.story_contents.get_node_with_corresponding_exp_id(
+                    exp_id))
+            old_node = (
+                old_story.story_contents.get_node_with_corresponding_exp_id(
+                    exp_id))
+            if old_node.title != new_node.title:
+                model_ids_need_update.add(exp_id)
+
+    update_exploration_opportunities_with_new_story(
+        new_story, list(model_ids_need_update))
+    add_new_exploration_opprtunities(new_story, new_added_exp_ids)
+    delete_exploration_opportunities(list(deleted_exp_ids))
+
+
+def get_translation_opportunities(language_code, cursor):
+    """Returns a list of opportunities available for translation in a specific
+    language.
+
+        Args:
+            cursor: str or None. If provided, the list of returned entities
+                starts from this datastore cursor. Otherwise, the returned
+                entities start from the beginning of the full list of entities.
+            language_code: str. The language for which translation opportunities
+                to be fetched.
+
+        Returns:
+            3-tuple(results, cursor, more). where:
+                results: list(dict). A list of dict of opportunity details.
+                cursor: str or None. A query cursor pointing to the next
+                    batch of results. If there are no more results, this might
+                    be None.
+                more: bool. If True, there are (probably) more results after
+                    this batch. If False, there are no further results after
+                    this batch.
+    """
+    page_size = feconf.OPPORTUNITIES_PAGE_SIZE
+    exp_opportunity_models, cursor, more = (
+        opportunity_models.ExplorationOpportunitySummaryModel.get_all_translation_opportunities( # pylint: disable=line-too-long
+            page_size, cursor, language_code))
+    results = []
+    for exp_opportunity_model in exp_opportunity_models:
+        results.append({
+            'topic': exp_opportunity_model.topic_name,
```

Comment:
I'm a bit worried about your choice of naming -- topic_name is different from topic, story_title is different from story. Maybe call them topic_name, story_title, etc.